商品購入機能の作成（サーバーサイド）

truthconductor opened the pull request:
# What
他のユーザーが出品した商品を購入する機能を追加する

# Why
商品購入のやりとりを行うため

# 補足
## テーブル追加
- delivery_addressesテーブル
配送先住所を記録するテーブル
- paymentsテーブル
支払い金額と使用ポイントを記録するテーブル
## テスト
単体テストは以下のコマンドを実行してください
bundle exec rake db:seed RAILS_ENV=test
bundle exec rspec spec/models/delivery_address_spec.rb
bundle exec rspec spec/models/deal_spec.rb

truthconductor committed:
データベースの商品情報を商品購入確認ページに反映

truthconductor committed:
発送先の取得、登録、変更機能追加

truthconductor committed:
バリデーションのエラーメッセージを表示（カラム名日本語化前）

truthconductor committed:
バリデーションメッセージの日本語化

truthconductor committed:
DeliveryAddressモデルの単体テストを追加

truthconductor committed:
phonelibを使用した電話番号バリデーションとテストの追加

truthconductor committed:
商品の購入処理を実装

商品状態や不正なリクエストで購入できない場合は商品詳細ページへリダイレクト、また購入ボタンの見た目をグレー化する
payjpのcard_tokenが取得できない時に、エラーにせずにリダイレクトするように変更
payjpのAPI設定をbefore_actionで行うように変更

truthconductor committed:
Dealモデルの単体テストを追加

truthconductor committed:
Merge branch 'master' into add-deals-new-serverside

truthconductor committed:
不要な記述の消去、データベース制約修正

不要なbinding.pryの記述を消去
paymentsテーブルのdeal_idのnullを禁止
取引テストの販売者、購入者の指定ミスを修正

truthconductor committed:
不要な記述の消去、データベース制約修正

不要なbinding.pryの記述を消去
paymentsテーブルのdeal_idのnullを禁止
取引テストの販売者、購入者の指定ミスを修正

Hiro0914 commented on app/views/deals/purchase/new.html.haml:
```diff
@@ -1,60 +1,72 @@
+              - else
+                .deals-content__section__w320__method__content
+                  カードを登録してください
+                = link_to new_item_purchase_card_path, class: "deals-content__section__w320__method__change" do
```
ターボリンクスが起動しているとページ読み込みを起点とした、
画面遷移の場合、javascriptが機能しません。
これではカード登録画面に遷移後、一度画面を更新しないと、
登録できないのでturbolinksはoffにしましょう。

= link_to new_item_purchase_card_path, class: "deals-content__section__w320__method__change", data: {"turbolinks":  false} do

truthconductor replied:
data: {"turbolinks": false}をlint_toの引数に指定し、
turbolinksを使わずにページ移動されるようにしました。

Hiro0914 replied:
マイページからのカード登録画面は修正確認しました。
購入画面からカード登録画面へのlink_toの記述が直っていません。
再度確認願います。

truthconductor replied:
すみません。
数行上のカード情報未登録側でしかリンク修正していませんでした。
カート情報登録済済側のリンクも修正しました。

Hiro0914 replied:
修正確認しました。

Hiro0914 commented on app/views/delivery_addresses/_error_message.html.haml:
```diff
@@ -0,0 +1,3 @@
+- full_messages.each do |full_message|
+  .validation-error__text
+    = full_message
```
未入力の際にエラーメッセージを全て出す必要はあるのでしょうか？
例えば名前をカナ入力するフォームが未入力であった場合、
①もしくは②どちらかのみで良い気もします。

①未入力に対するエラーメッセージ
お名前カナ（姓）を入力してください
②フォーマットに対するエラーメッセージ
お名前カナ（姓）を全角カナで入力してください

truthconductor replied:
未入力のバリデーションエラーが存在するときは、
最初のバリデーションエラーを表示するように変更しました。
モデルのバリデーション定義で未入力エラーを最初に記述する事で、
未入力の場合のエラーメッセージのみを表示させるようにルール化しています。

Hiro0914 replied:
確認しました。

truthconductor committed:
Merge branch 'add-deals-new-serverside' of https://github.com/Hiro0914/freemarket_sample_54b into add-deals-new-serverside

Hiro0914 commented on spec/models/delivery_address_spec.rb:
```diff
@@ -0,0 +1,257 @@
+        expect(delivery_address.errors[:first_name_kana]).to include("を入力してください")
+      end
+      it "is valid a first_name_kana that has less than or equal to 35 characters" do
+        delivery_address = build(:delivery_address, first_name_kana:"アイウエオカキクケコアイウエオカキクケコアイウエオカキクケコアイウエオ")
```
テストコードの条件となる文字列は、毎回同じではなく、
以下のようにコードで作成してランダムにすべきかと思います。

ex.) ランダムなカタカナ35文字の文字列を作成
('ァ'..'ヴ').to_a.sample(35).join

truthconductor replied:
氏名入力は
Faker::String.random(length: 35)
カナ入力は
('ァ'..'ヴ').to_a.sample(35).join
を使用しランダムな文字列でテストするように変更しました。

Hiro0914 replied:
確認しました。

truthconductor committed:
リンク移動・エラー表示・テスト記述の修正

payjpのjavascriptが動作するページのリンクでturbolinksが動作しないようにした。
バリデーションエラーメッセージで入力が空欄のエラーが発生したときは、他のバリデーションエラーを表示させないようにした。
文字数のテストを固定文字ではなく、ランダム生成された文字を使用するようにした。

Hiro0914 commented:
すみません、１つ追加です。
カードが既に登録された状態で購入画面へ遷移、
”支払方法を変更する”をクリック後に表示されるクレジットカード一覧のページについてです。
この画面ですが、カードを複数枚登録させないのであれば、
最初からラジオボタンをオンにした方が良いかと思います。
(本家もオンになってます。まあそもそもボタンが必要なのか微妙な気もしますが。)

ラジオボタンの選択制御は、jqueryのpropメソッドで行えます。

truthconductor commented:
jqueryを使用しなくてもフォームヘルパーで選択指定が可能なので、その方法で修正しました。

本家はカード変更と書かれていながら、カードが１枚しか使用できなかったりと中途半端な状態なので、
必須機能開発後に時間が取れれば複数枚カードの登録・選択もできればとは思ってます。
カード情報はpayjp側に複数枚登録されていれば複数枚取得できるように考慮していますが、
それを元にしてデフォルトカードを切り替えする機能は本家同様現時点では行っていないです。

truthconductor committed:
turbolinks無効化抜け修正、カード選択ラジオボタンのデフォルトチェック追加

payjpのjavascriptが動作するページのリンクでturbolinksが動作しないようする箇所の設定抜けを修正。
カード選択の際に

Hiro0914 commented:
LGTM

TOMARUMARU commented on app/controllers/deals/purchase_controller.rb:
```diff
@@ -2,12 +2,74 @@ class Deals::PurchaseController < ApplicationController
   def new
     # 購入確認商品を取得する
-    @item = Item.find(params[:item_id])
+    @item = Item.find(params[:item_id]).decorate
```
他のアクションでも使用しているのでメソッドに切り分けましょう。

truthconductor replied:
get_itemメソッドを作成し、
before_actionで呼び出しを行うようにしました。

TOMARUMARU commented on app/controllers/deals/purchase_controller.rb:
```diff
@@ -2,12 +2,74 @@ class Deals::PurchaseController < ApplicationController
+    unless @item.sales_state_id == 1
+      redirect_to item_path(id: @item.id)
+      return
+    end
```
後置unlessで1行で書けそうですね

truthconductor replied:
後置unlessを使用したワンライナーに書き換えました。
createメソッド内でも同様に書ける箇所があったため、そちらも合わせて修正しました。

TOMARUMARU commented on app/models/credit_card.rb:
```diff
@@ -4,8 +4,7 @@ class CreditCard < ApplicationRecord
   def getPayJPCards()
-    require 'payjp'
-    Payjp.api_key = Rails.application.credentials.payjp[:api_key]
+    set_api_key
```
before_actionで呼びましょう

truthconductor replied:
ApplicationRecordを継承したCreditCardクラスのため、
before_actionは使わず（ApplicationControllerを継承しておらず使えないので）
after_initializeコールバックでAPIキーを初期化するようにしました。

truthconductor committed:
共通処理のコールバック呼び出し化、ワンライナー記述への修正

PurchaseControllerのリダイレクト早期リターン処理をワンライナー化
PurchaseControllerの商品情報取得をbefore_actionによるコールバック呼び出しで行うように変更
CreditCardモデル内のpayjp.APIキーセットを、after_initializeコールバック内で行いインスタンス化された常に行われるように変更

TOMARUMARU commented on app/controllers/deals/purchase_controller.rb:
```diff
@@ -2,12 +2,67 @@ class Deals::PurchaseController < ApplicationController
-    # 購入確認商品を取得する
-    @item = Item.find(params[:item_id])
+    # 商品が出品中でないのにリクエストが来た時商品詳細ページへリダイレクト
+    return redirect_to item_path(id: @item.id) unless @item.sales_state_id == 1
```
item_path(@item)で大丈夫です！他の箇所も同様です

truthconductor replied:
returnを用いないとその後に続くメソッドの処理が実行されてしまいませんか？
購入処理を止めずに行ってしまったりレンダリングエラーが発生する恐れがあるので、returnは外さない方が良いと思うのですが、外す理由がほかにあるのでしょうか？
 
参考記事
http://dev.trick-with.net/2012/12/redirect_to-and-return-is-correct-or-not/
https://railsguides.jp/layouts_and_rendering.html#redirect-toを使用する
--引用--
redirect_toやredirect_backはメソッドの実行を即座に中断したりはしません。これらは単にHTTPのレスポンスを設定するだけです。もしこれらの後にメソッドがあった場合そのメソッドは実行されてしまいます。必要であれば、明示的なreturnもしくは他の中断用の手法を使うことで中断可能です。

TOMARUMARU replied:
指摘してるのは引数の話です！

truthconductor replied:
指摘が理解できておらず失礼しました。
修正しましたらまたレビュー出します。

truthconductor replied:
引数を@itemに変更しました

truthconductor committed:
pathヘルパーの引数でモデルを引数に渡すように修正、パス指定の引数指定間違いを修正

ken-nag commented on spec/models/delivery_address_spec.rb:
```diff
@@ -0,0 +1,257 @@
+        delivery_address = build(:delivery_address, prefecture_id:Faker::Number.between(from: 1, to: 47))
+        delivery_address.user = build(:user)
+        expect(delivery_address).to be_valid
+      end
```
it~endの下に改行欲しいですね、可読性のために

truthconductor replied:
テスト毎に改行を加えました。

truthconductor committed:
単体テストコード間に１行空きを追加

ken-nag commented on spec/models/deal_spec.rb:
```diff
@@ -0,0 +1,59 @@
+      deal = build(:deal, buyer_id:2, seller_id:nil)
+      deal.item = build(:item)
+      deal.valid?
+      expect(deal.errors[:seller]).to include("を入力してください")
```
letなどで共通化できる部分が多いので検討しましょう
deal = build(:deal, buyer_id:2, seller_id:nil)
deal.item = build(:item)
deal.valid?
ここは特に使用可能なのでsubjectなども用いて簡潔に書けると思います

truthconductor replied:
letでそれぞれのモデルを作成するように定義し、
subjectでバリデーションの検証をまとめました。

ken-nag commented on app/models/credit_card.rb:
```diff
@@ -2,10 +2,15 @@ class CreditCard < ApplicationRecord
 
+  # インスタンス化された際に実行
+  after_initialize do
+    require "payjp"
```
ここのrequire不要だと思います

truthconductor replied:
requireを消去しました

truthconductor committed:
Dealsモデル作成の単体テストにletとsubjectを適用

shingoteshima commented:
ＬＧＴＭ
